ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,167 @@
+    | ((days_curr, _) as con_curr) :: cons, None ->
+        if remaining_days <= float_of_int days_curr then
+          get_most_critical (cons, Some con_curr)
+        else
+          get_most_critical (cons, None)
+    | ( ((days_curr, _) as con_curr) :: cons
+      , (Some (days_critical, _) as con_critical) ) ->
+        if
+          remaining_days <= float_of_int days_curr && days_curr <= days_critical
+        then
+          get_most_critical (cons, Some con_curr)
+        else
+          get_most_critical (cons, con_critical)
+  in
+  let critical_condition =
+    get_most_critical (message_sent_on_remaining_days_list, None)
+  in
+  match critical_condition with
+  | None ->
+      None
+  | Some (days, expired_message_id) when days <= 0 ->
+      let expired = expired_message obj_description in
+      Some (message_body expired expiry, expired_message_id)
+  | Some (_, expiring_message_id) ->
+      let expiring = expiring_message obj_description in
+      Some (message_body expiring expiry, expiring_message_id)
+
+let messages_require_update alert related_msgs =
+  let msg_body, (msg_name, msg_prio) = alert in
+  let is_outdated (_ref, record) =
+    record.API.message_body <> msg_body
+    || record.API.message_name <> msg_name
+    || record.API.message_priority <> msg_prio
+  in
+  let outdated, current = List.partition is_outdated related_msgs in
+  (outdated, current = [])
+
+let update_message_internal message_name_list msg_obj_uuid alert all_msgs =
+  all_msgs
+  |> related_messages msg_obj_uuid message_name_list
+  |> messages_require_update alert
+
+let update_message rpc session_id msg_name_list msg_cls msg_obj_uuid alert
+    all_msgs =
+  let outdated_msgs, create_new_msg =
+    all_msgs |> update_message_internal msg_name_list msg_obj_uuid alert
+  in
+  List.iter
+    (fun (self, _) -> XenAPI.Message.destroy ~rpc ~session_id ~self)
+    outdated_msgs ;
+  if create_new_msg then
+    let body, (name, priority) = alert in
+    XenAPI.Message.create ~rpc ~session_id ~name ~priority ~cls:msg_cls
+      ~obj_uuid:msg_obj_uuid ~body
+    |> ignore
+
+let alert ~rpc ~session_id expiry_messaging_info_list =
+  let now = Date.now () in
+  let alert_message_info_list =
+    List.filter_map
+      (fun {
+             message_cls
+           ; message_obj_uuid
+           ; obj_description
+           ; message_sent_on_remaining_days_list
+           ; expiry
+           } ->
+        let alert =
+          generate_alert now obj_description message_sent_on_remaining_days_list
+            expiry
+        in
+        match alert with
+        | Some alert ->
+            let msg_name_list =
+              List.map
+                (fun (_, (msg_name, _)) -> msg_name)
+                message_sent_on_remaining_days_list
+            in
+            Some (alert, msg_name_list, message_cls, message_obj_uuid)
+        | None ->
+            None
+      )
+      expiry_messaging_info_list
+  in
+  if alert_message_info_list <> [] then
```

Comment:
I think this is single data flow, the processing can be appended simply. For example:
```
expiry_messaging_info_list
|> List.filter_map (...)
|> List.iter (...)
```

Reply:
`all_msgs` needs to be retrieved before List.iter only if List.filter_map returns an non-empty list.

Reply:
Then `all_msgs` will always be retrieved 1 time no matter if there is any alert required to be updated?